ttnn/cpp/ttnn/operations/creation.hpp

```diff
@@ -218,14 +218,26 @@ inline constexpr ZerosLike zeros_like{};
 inline constexpr OnesLike ones_like{};
 
 struct Empty {
-   static ttnn::Tensor invoke(
-    const ttnn::Shape& shape,
-    const DataType& dtype,
-    const Layout& layout,
-    Device* device,
-    const MemoryConfig& memory_config) {
+    static ttnn::Tensor invoke(
+        const ttnn::Shape& shape,
+        const DataType& dtype,
+        const Layout& layout,
+        Device* device,
+        const MemoryConfig& memory_config) {
```

Comment:
@patrickroberts since I saw you recently touched these creation functions. fyi @eyonland @yan-zaretskiy 

```
constexpr auto full =
    ttnn::decorators::register_operation_with_auto_launch_op<"ttnn::full", ttnn::operations::creation::Full>();
constexpr auto zeros = ttnn::decorators::register_operation<"ttnn::zeros", ttnn::operations::creation::Zeros>();
constexpr auto ones = ttnn::decorators::register_operation<"ttnn::ones", ttnn::operations::creation::Ones>();
constexpr auto empty = ttnn::decorators::register_operation<"ttnn::empty", ttnn::operations::creation::Empty>();

constexpr auto full_like =
    ttnn::decorators::register_operation_with_auto_launch_op<"ttnn::full_like", ttnn::operations::creation::FullLike>();
constexpr auto zeros_like =
    ttnn::decorators::register_operation<"ttnn::zeros_like", ttnn::operations::creation::ZerosLike>();
constexpr auto ones_like =
    ttnn::decorators::register_operation<"ttnn::ones_like", ttnn::operations::creation::OnesLike>();
constexpr auto empty_like =
    ttnn::decorators::register_operation<"ttnn::empty_like", ttnn::operations::creation::EmptyLike>();

constexpr auto arange =
    ttnn::decorators::register_operation_with_auto_launch_op<"ttnn::arange", ttnn::operations::creation::Arange>();

```

Some of these creation functions use `register_operation_with_auto_launch_op` and some use `register_operation`. For the ones that use `register_operation` I don't see see explicit calls to `launch_op`. Do we want to make all these invoke functions for these creation ops to be done under the scope of launch_op? It seems inconsistent right now?

Reply:
When I updated those creation functions, I did not change whether they opted into `auto_launch_op`, that was determined by @arakhmati, and I don't have a good understanding of the criteria he used to decide that.

Reply:
When op returns something else than a `Tensor`, for example a vector of tensors. In this case infra can't determine how many tensors to create or which to create so automatic decoration is not possible. In this case you must use `register_operation` and make an explicit call to launch_op inside operation's invoke.

If this is not handled correctly, operation won't be async even if async mode is enabled.

Good news, I am sure 6 months from now this wonderful infra should go away.

Reply:
https://github.com/tenstorrent/tt-metal/pull/10672 introduces the discrepancy, I think we can fix in a follow up? Under the hood, full/zeros/ones are ultimately the same; empty path is a bit simpler as it only requires an allocation.

Reply:
For now, your changes now make the single-device and multi-device operation work with async. It's also consistent between the two so looks good for now.